File: android/app/src/main/kotlin/run/nuri/getagrip/ui/components/Chips.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import run.nuri.getagrip.engine.GripPosition
import run.nuri.getagrip.engine.HandMode
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.GetAGripTheme
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics

// The app's ONE selection control, ported from Sources/UI/Components/Chips.swift.
//
// Chips survive only where the choice is CATEGORICAL (position, hand mode, sessions a day,
// ritual-or-whenever). Every QUANTITY moved to `ValueRow` (Nuri, 2026-08-03): nine edge
// chips still cannot express 22 mm.

/// One capsule option.
///
/// Selected is a tonal graphite wash; unselected a hairline capsule with no fill. The
/// asymmetry is load-bearing: fill-vs-outline survives greyscale, Reduce Transparency and
/// colourblindness, where a tint step does not.
///
/// TRANSLATION NOTE: iOS limits glass for re-blur cost across twenty-plus chips; Android has
/// no glass, so the tonal wash IS the surface.
@Composable
fun Chip(
    title: String,
    isSelected: Boolean,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val palette = LocalGripPalette.current
    val interactionSource = remember { MutableInteractionSource() }
    Surface(
        // Drawn at 44 and SHAPED at 44: the visual IS the target.
        modifier = modifier
            .heightIn(min = chipMinHeight)
            .semantics {
                role = Role.Button
                selected = isSelected
            }
            .clickable(
                enabled = enabled,
                interactionSource = interactionSource,
                indication = null,
                onClick = onClick,
            )
            .pressFeedback(interactionSource),
        shape = CircleShape,
        color = if (isSelected) palette.graphite.copy(alpha = 0.20f) else Color.Transparent,
        border = if (isSelected) null else BorderStroke(1.dp, palette.inkTertiary.copy(alpha = 0.35f)),
    ) {
        Row(
            Modifier.padding(horizontal = Metrics.buttonHorizontalPadding, vertical = Metrics.buttonVerticalPadding),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                title,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Medium,
                color = if (isSelected) palette.inkPrimary else palette.inkSecondary,
                textAlign = TextAlign.Center,
            )
        }
    }
}

/// The house hit-target floor; the preset row once sat four points under it and taps read as
/// unregistered. (Not `CHIP_HEIGHT`: `DeviceChip` owns that for its 40 dp drawn pill.)
val chipMinHeight: Dp = 44.dp

// MARK: - Layout

/// The shared chip layout: cells of at least `minimumChipWidth`, sharing the row equally.
///
/// **Chip density is content-driven, never a fixed column count.** A six-column grid
/// squeezed cells to ~48 dp and the edge row read "10…, 12…, 15…, 18…" — four edges made
/// indistinguishable.
///
/// TRANSLATION NOTE: SwiftUI's `LazyVGrid(.adaptive(minimum:))`, reproduced with
/// `BoxWithConstraints` and eager rows: a handful of chips costs nothing to lay out, and a lazy
/// grid inside the builder's scrolling document would need a fixed height.
@Composable
fun ChipGrid(
    base: Int,
    modifier: Modifier = Modifier,
    spacing: Dp = 8.dp,
    content: List<@Composable (Modifier) -> Unit>,
) {
    val minimum = minimumChipWidth(base)
    BoxWithConstraints(modifier.fillMaxWidth()) {
        val available = maxWidth
        val columns = chipColumns(available, minimum, spacing).coerceAtMost(maxOf(1, content.size))
        Column(verticalArrangement = Arrangement.spacedBy(spacing)) {
            content.chunked(columns).forEach { row ->
                Row(Modifier.height(IntrinsicSize.Min), horizontalArrangement = Arrangement.spacedBy(spacing)) {
                    row.forEach { cell -> cell(Modifier.weight(1f).fillMaxHeight()) }
                    // The last row keeps the column WIDTH: chips changing size between lines read as two controls.
                    repeat(columns - row.size) { Spacer(Modifier.weight(1f)) }
                }
            }
        }
    }
}

/// Cells of at least `minimum` that fit. At least one: a phone narrower than a chip still draws it.
internal fun chipColumns(available: Dp, minimum: Dp, spacing: Dp): Int {
    if (available <= 0.dp) return 1
    val n = ((available + spacing).value / (minimum + spacing).value).toInt()
    return maxOf(1, n)
}

/// `base` is the row's column count at standard text: numeric rows (6) want narrow cells,
/// "Half crimp" (3) and "Alternate each pull" (2) wide ones. It sets a MINIMUM width, so the
/// row wraps a chip rather than truncating one.
///
/// TRANSLATION NOTE: iOS steps at `.accessibility1` and `.accessibility3`; font scales 1.3
/// and 1.6 land in the same places.
@Composable
private fun minimumChipWidth(base: Int): Dp {
    val standard: Dp = when {
        base <= 2 -> 150.dp
        base == 3 -> 104.dp
        base <= 5 -> 76.dp
        else -> 56.dp
    }
    val scale = LocalDensity.current.fontScale
    return when {
        scale >= 1.6f -> standard * 2.2f
        scale >= 1.3f -> standard * 1.6f
        else -> standard
    }
}

// MARK: - Whole-number rows

/// A small categorical integer choice, currently sessions per day. Preserve an off-menu
/// stored value as its own chip rather than changing a draft merely by opening it.
@Composable
fun IntChipRow(
    values: List<Int>,
    selection: Int,
    modifier: Modifier = Modifier,
    unit: String = "",
    onSelect: (Int) -> Unit,
) {
    val haptics = LocalHapticFeedback.current
    fun label(value: Int) = if (unit.isEmpty()) "$value" else L10n.tr("%d %s", value, unit)
    val drawn = if (values.contains(selection)) values else (values + selection).sorted()
    val longest = drawn.maxOfOrNull { label(it).length } ?: 1
    val density = when {
        longest <= 2 -> 6
        longest <= 4 -> 5
        else -> 4
    }
    ChipGrid(
        base = density,
        modifier = modifier,
        content = drawn.map { value ->
            { cellModifier: Modifier ->
                Chip(label(value), value == selection, cellModifier) {
                    if (value != selection) {
                        haptics.performHapticFeedback(HapticFeedbackType.SegmentTick)
                        onSelect(value)
                    }
                }
            }
        },
    )
}

/// How the hand is SET on the edge — a genuinely categorical choice, so it stays chips.
@Composable
fun PositionChipRow(
    selection: GripPosition,
    modifier: Modifier = Modifier,
    onSelect: (GripPosition) -> Unit,
) {
    val haptics = LocalHapticFeedback.current

    /// A position from a newer build still gets a (selected) chip — why `GripPosition` is an
    /// open value class. Dropping it would silently rewrite the set.
    val options = if (GripPosition.known.contains(selection)) {
        GripPosition.known
    } else {
        GripPosition.known + selection
    }

    fun choose(position: GripPosition) {
        if (position == selection) return
        haptics.performHapticFeedback(HapticFeedbackType.SegmentTick)
        onSelect(position)
    }

    Column(modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        ChipGrid(
            base = 3,
            content = options.map { position ->
                { cellModifier: Modifier ->
                    Chip(position.name, position == selection, cellModifier) { choose(position) }
                }
            },
        )
        if (selection == GripPosition.fingerCurl) {
            Text(
                tr("Start in half crimp and curl your fingers into the edge."),
                style = MaterialTheme.typography.bodySmall,
                color = LocalGripPalette.current.inkSecondary,
            )
        }
    }
}

/// How a set is shared between hands.
@Composable
fun HandModeChipRow(
    selection: HandMode,
    modifier: Modifier = Modifier,
    onSelect: (HandMode) -> Unit,
) {
    val haptics = LocalHapticFeedback.current
    // Two across at most: "Alternate each pull" cannot be read in a narrow cell.
    ChipGrid(
        base = 2,
        modifier = modifier,
        content = HandMode.entries.map { mode ->
            { cellModifier: Modifier ->
                Chip(mode.displayName, mode == selection, cellModifier) {
                    if (mode != selection) {
                        haptics.performHapticFeedback(HapticFeedbackType.SegmentTick)
                        onSelect(mode)
                    }
                }
            }
        },
    )
}

@Preview(name = "Chips", showBackground = true, widthDp = 380)
@Composable
private fun ChipsPreview() {
    GetAGripTheme {
        var sessions by remember { mutableStateOf(2) }
        var position by remember { mutableStateOf(GripPosition.halfCrimp) }
        var hands by remember { mutableStateOf(HandMode.alternateEachRep) }
        Column(
            Modifier.padding(20.dp).width(340.dp),
            verticalArrangement = Arrangement.spacedBy(18.dp),
        ) {
            CapsLabel(tr("Sessions a day"))
            IntChipRow(listOf(1, 2, 3, 4), sessions) { sessions = it }
            CapsLabel(tr("GRIP"))
            PositionChipRow(position) { position = it }
            CapsLabel(tr("HANDS"))
            HandModeChipRow(hands) { hands = it }
        }
    }
}

File: Sources/UI/Components/Chips.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

// Chips express categorical choices: position, hands and sessions per day.
// Quantities use ValueRow, where a slider and direct entry share the same value.

/// One capsule option.
///
/// Selected draws a single glass surface; unselected is a hairline capsule with **no
/// glass at all**. Load-bearing: an expanded set row can carry twenty-plus chips, and
/// every glass surface re-blurs its backdrop.
struct Chip: View {
    var title: String
    var isSelected: Bool
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title)
                .font(.system(.subheadline, weight: isSelected ? .semibold : .medium))
                .monospacedDigit()
                .foregroundStyle(isSelected ? Ink.primary : Ink.secondary)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.horizontal, 12)
                .padding(.vertical, Metrics.buttonVerticalPadding)
                // Drawn at 44 and shaped at 44: the visual IS the target, and padding plus
                // a capsule background are not hit-tested without the content shape.
                .frame(maxWidth: .infinity, minHeight: 44)
                .modifier(ChipSurface(isSelected: isSelected))
                .contentShape(.capsule)
        }
        .buttonStyle(PressFeedbackButtonStyle())
        .accessibilityAddTraits(isSelected ? [.isSelected] : [])
    }
}

/// The two chip states, kept in one place so no row can invent a third.
private struct ChipSurface: ViewModifier {
    var isSelected: Bool

    func body(content: Content) -> some View {
        if isSelected {
            // `.accessibleGlass`: under Reduce Transparency raw glass loses its tint,
            // the only thing saying which chip is chosen.
            content.accessibleGlass(Accent.graphiteFlat.opacity(0.20), in: .capsule)
        } else {
            content.overlay(Capsule().stroke(Ink.tertiary.opacity(0.35), lineWidth: 1))
        }
    }
}

// MARK: - Layout

/// The shared chip layout: six across, thinning as text grows (6 → 3 at
/// `.accessibility1` → 2 at `.accessibility3`). `base` is the comfortable count for
/// THIS row's labels.
struct ChipGrid<Content: View>: View {
    var base: Int
    @ViewBuilder var content: Content
    @Environment(\.dynamicTypeSize) private var typeSize

    var body: some View {
        LazyVGrid(columns: columns, alignment: .leading, spacing: 8) {
            content
        }
    }

    private var columns: [GridItem] {
        // ADAPTIVE, not a fixed count: a fixed 6-column grid squeezed cells to ~48pt
        // and the edge row rendered "6…, 8…, 1…, 1…, 1…, 1…" — 10, 12, 15 and 18 mm
        // indistinguishable. `minimum` scales with type size, so the 6 → 3 → 2 ladder
        // still happens; rows with a unit take one fewer column instead of truncating.
        [GridItem(.adaptive(minimum: minimumChipWidth), spacing: 8, alignment: .leading)]
    }

    /// `base` is the row's intended column count at standard type (6 for numbers, 3 for
    /// "Half crimp", 2 for "Alternate each pull"). It sets a minimum width rather than a hard
    /// count, so the row wraps a chip instead of truncating one.
    private var minimumChipWidth: CGFloat {
        let standard: CGFloat = switch base {
        case ...2: 150
        case 3: 104
        case 4...5: 76
        default: 56
        }
        if typeSize >= .accessibility3 { return standard * 2.2 }
        if typeSize >= .accessibility1 { return standard * 1.6 }
        return standard
    }
}

// MARK: - Whole-number rows

/// A small categorical integer choice, currently sessions per day. An off-menu stored
/// value receives its own chip so merely opening the editor never rewrites it.
struct IntChipRow: View {
    let values: [Int]
    var unit: String = ""
    @Binding var selection: Int

    private var drawnValues: [Int] {
        values.contains(selection) ? values : (values + [selection]).sorted()
    }

    private var densityForLongestLabel: Int {
        switch drawnValues.map({ label($0).count }).max() ?? 1 {
        case ...2: 6
        case 3...4: 5
        default: 4
        }
    }

    var body: some View {
        ChipGrid(base: densityForLongestLabel) {
            ForEach(drawnValues, id: \.self) { value in
                Chip(title: label(value), isSelected: value == selection) { selection = value }
            }
        }
        .sensoryFeedback(.selection, trigger: selection)
    }

    private func label(_ value: Int) -> String {
        unit.isEmpty ? "\(value)" : "\(value) \(unit)"
    }
}

// MARK: - Grip position

struct PositionChipRow: View {
    @Binding var selection: GripPosition
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            ChipGrid(base: 3) { chips }
            if selection == .fingerCurl {
                Text("Start in half crimp and curl your fingers into the edge.")
                    .font(.system(.caption))
                    .foregroundStyle(Ink.secondary)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        .sensoryFeedback(.selection, trigger: selection)
    }

    @ViewBuilder
    private var chips: some View {
        ForEach(options, id: \.rawValue) { position in
            Chip(title: position.name, isSelected: position == selection) {
                selection = position
            }
        }
    }

    /// A position written by a newer build is not in `known` and still gets a chip — the
    /// reason `GripPosition` is an open struct. Dropping it would let this build silently
    /// rewrite the set.
    private var options: [GripPosition] {
        GripPosition.known.contains(selection) ? GripPosition.known : GripPosition.known + [selection]
    }
}

/// How a set is shared between hands.
struct HandModeChipRow: View {
    @Binding var selection: HandMode

    // Two across at most: "Alternate each pull" cannot be read in a 48pt cell.
    var body: some View {
        ChipGrid(base: 2) {
            ForEach(HandMode.allCases, id: \.self) { mode in
                Chip(title: mode.name, isSelected: mode == selection) {
                    selection = mode
                }
            }
        }
        .sensoryFeedback(.selection, trigger: selection)
    }
}

/// Which hand a max was pulled with.
///
/// **"Both hands" leads and is the default**: the honest answer for anyone who has not
/// thought about it, and the pre-hands behaviour (applies to every rep of that grip).
/// Choosing a side says "this number is NOT true of my other hand", and the engine treats
/// it that way — see `MaxTable`.
struct MaxSideChipRow: View {
    @Binding var selection: Side

    /// Both first, then the hands in the order the runner alternates them.
    private static let order: [Side] = [.both, .left, .right]

    var body: some View {
        ChipGrid(base: 3) {
            ForEach(Self.order, id: \.self) { side in
                Chip(title: name(side), isSelected: selection == side) {
                    selection = side
                }
                .accessibilityLabel(spoken(side))
            }
        }
        .sensoryFeedback(.selection, trigger: selection)
    }

    /// NOT `Side.name`, the runner's arm's-length vocabulary ("Left", "Both"). Here the
    /// question is which hand the max is for, answered by "Both hands", "Left hand".
    private func name(_ side: Side) -> String {
        switch side {
        case .both:  String(localized: "Both hands")
        case .left:  String(localized: "Left hand")
        case .right: String(localized: "Right hand")
        }
    }

    private func spoken(_ side: Side) -> String {
        side == .both ? String(localized: "For both hands") : String(localized: "For the \(side.name.lowercased()) hand only")
    }
}

#Preview {
    @Previewable @State var hold = 10
    @Previewable @State var position = GripPosition.halfCrimp
    @Previewable @State var hands = HandMode.alternateEachRep
    @Previewable @State var side = Side.both

    ScrollView {
        VStack(alignment: .leading, spacing: 18) {
            CapsLabel("SESSIONS A DAY")
            IntChipRow(values: [1, 2, 3, 4], selection: $hold)
            CapsLabel("GRIP")
            PositionChipRow(selection: $position)
            CapsLabel("HANDS")
            HandModeChipRow(selection: $hands)
            CapsLabel("THIS MAX IS FOR")
            MaxSideChipRow(selection: $side)
        }
        .padding(Metrics.hPadding)
    }
    .background { AppBackground() }
}
